rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,218 @@
+- Start Date: 2020-06-02
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+Pluggable state from registry items can often end up stored inside saved objects, for instance:
+Embeddables, Expression strings, Url generators, ...
+
+When plugin A (persister) stores some state that belongs to another plugin B a few issues arise:
+- How does persister know if plugin B state contains references to saved objects
+- How does persister migrate the saved object when it contains state that belongs to plugin B
+- How does persister know if state that belongs to plugin B contains state that belongs to plugin C
+
+# Basic example
+
+```ts
+
+const MY_STATE_ID = 'MyState';
+
+interface MyStateV76 extends PersistableState {
+  object: string,
+  val: number,
+}
+
+interface MyStateV77 extends PersistableState {
+  objectId: string,
+  val: number,
+}
+
+interface MyState extends PersistableState {
+  objectId: string,
+  value: number,
+}
+
+const migrate = (state: unknown, version: string) => {
+  let v76 = version === '7.6' ? state as MyStateV76 : undefined;
+  let v77 = version === '7.7' ? state as MyStateV77 : undefined;
+  let v78 = version === '7.8' ? state as MyState : undefined;
+  if (v76) {
+    v77 = { objectId: v76.object, val: v76.val };
+  }
+  if (v77) {
+    v78 = { objectId: v77.objectId, value: v77.val };
+  }
+  
+  return newState;
+}
+
+const inject = (state: MyState, savedObjectReferences: SavedObjectReference[]) => {
+  return {
+    ...state,
+    objectId: savedObjectReferences.find(ref => ref.name = state.objectId)?.id;
+  }
+}
+
+const extract = (state: MyState) => {
+  const references = [{ name: 'objectId', id: state.objectId, type: 'savedObject' }];
+  return [{ ...state, objectId: 'objectId' }, references];
+}
+
+persistableStateService.register(MY_STATE_ID, { migrate, inject, extract });
+```
+
+```ts
+  const stateReadyForPersisting = persistableStateService.get(MY_STATE_ID),beforeSave(myState);
+  const stateReadyForConsuming = persistableStateService.get(MY_STATE_ID).afterLoad(stateReadyForPersisting);
```

Comment:
Maybe would be clearer to show an example `SavedObjects.get(...)` and then using that to consume the state? It's slightly confusing to have `stateReadyForPersisting` included again here

Reply:
Also, use `persistableStateService.get(id)` not `persistableStateService.get(MY_STATE_ID)` to indicate the id is not known at compile time?